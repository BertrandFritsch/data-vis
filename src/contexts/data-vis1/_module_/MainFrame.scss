
.data-vis-container {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 1000px;
  margin: 0 auto;

  .data-vis-header {
    margin-top: 20px;
    display: flex;
    flex-direction: column;
    p {
      text-align: center;
    }

    .data-vis-header-1 {
      font-size: 2rem;
    }

    .data-vis-header-2 {
      font-size: 1.5rem;
    }
  }

  .data-vis-body {
    flex: 1;
    display: flex;
    min-height: 500px;

    .data-vis-chart {
      flex: 1;
      align-self: stretch;
      margin: 100px 0 100px 100px;
      min-width: 100px;

      svg {

        .data-vis-line {
          fill: none;
          stroke-linejoin: round;
          stroke-linecap: round;
          stroke-width: 1.5;

          &.data-vis-line-over {
            stroke-width: 2.5;
          }
        }

        .data-vis-axis-y {
          .domain {
            display: none;
          }
          .tick line {
            stroke: #767676;
          }

          .tick:not(.data-vis-axis-y-zero):not(:first-of-type) line {
            stroke-dasharray: 2, 2;
          }

          .data-vis-axis-y-zero line {
            stroke-width: 2;
          }
        }

        text {
          font-size: .75rem;
          font-family: ibm-plex-sans;
        }
      }
    }

    .data-vis-tools {
      align-self: center;

      margin: 100px 50px;
      display: flex;
      flex-direction: column;

      .bx--toggle__label {
        width: 10rem;
        margin: .5rem;
        font-size: .875rem;

        .bx--toggle__appearance {
          font-size: .875rem;
          margin-right: .5rem;
        }
      }

      .data-vis-container-place-over {
        font-weight: 600;
      }
    }
  }

  .data-vis-footer {
    display: flex;
    flex-direction: column;
    margin: 10px;
    p {
      text-align: center;
      font-size: .75rem;
    }
  }
}
