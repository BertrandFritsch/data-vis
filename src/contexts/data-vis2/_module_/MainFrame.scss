
.data-vis-container {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 1000px;
  margin: 0 auto;

  .data-vis-header {
    margin-top: 20px;
    display: flex;
    flex-direction: column;
    p {
      text-align: center;
    }

    .data-vis-header-1 {
      font-size: 2rem;
    }

    .data-vis-header-2 {
      font-size: 1.5rem;
    }
  }

  .data-vis-body {
    flex: 1;
    display: flex;
    height: 500px;
    width: 500px;
    align-self: center;

    .data-vis-chart {
      flex: 1;
      align-self: stretch;
      margin: 20px;

      svg {

        circle {
          fill: #6898cf;

          &:hover {
            fill: #5b7faf;
          }
        }

        text {
          font-size: .2rem;
          font-family: ibm-plex-sans;
          text-anchor: middle;
          dominant-baseline: central;
          fill: white;
          pointer-events: none;
        }
      }
    }

    .data-vis-tooltip {
      opacity: 0;
      position: absolute;
      pointer-events: none;
      transition: opacity 300ms;

      &.data-vis-tooltip--hover {
        opacity: .9;
      }

      div {
        font-weight: 100;
        font-size: .75rem;
        background-color: black;
        color: white;
        text-align: left;
        padding: 10px;
        border: 0;
        border-radius: 2px;

        span {
          line-height: 1.25rem;
        }
      }
    }
  }

  .data-vis-footer {
    display: flex;
    flex-direction: column;
    margin: 10px;
    p {
      text-align: center;
      font-size: .75rem;
    }
  }
}
