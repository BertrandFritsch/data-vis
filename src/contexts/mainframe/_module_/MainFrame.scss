
.data-vis-container {
  display: flex;
  flex-direction: column;
  height: 100%;

  .data-vis-header {
    height: 50px;
  }

  .data-vis-body {
    flex: 1;
    padding: 100px;

    svg {
      height: 100%;
      width: 100%;
    }

  }
}