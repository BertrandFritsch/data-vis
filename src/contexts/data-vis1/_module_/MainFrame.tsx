import * as React from 'react';
import './MainFrame.scss';
import * as d3 from 'd3';
import { PlaceProperties, TemperatureByDate, TemperaturesByPlace, WorldPlace } from './reducers';
import { setInArray } from '../../../helpers';

export interface Props {
  data: TemperaturesByPlace[];
  minValue: number;
  maxValue: number;
  minDate: Date;
  maxDate: Date;
  placeProperties: PlaceProperties;
}

interface State {
  selectedPlaces: WorldPlace[];
}

interface ChartState {
  update: () => void;
  resize: () => void;
}

export default class MainFrame extends React.PureComponent<Props, State> {
  state: State;
  private chartState: ChartState = { update: () => undefined, resize: () => undefined };

  constructor(props: Props) {
    super(props);

    this.state = { selectedPlaces: props.data.find(d => d.place === 'Glob') ? [ 'Glob' ] : [] };
  }

  componentDidMount() {
    window.addEventListener('resize', this.windowResized);
  }

  componentDidUpdate() {
    this.chartState.update();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.windowResized);
  }

  render() {
    return (
      <div className='data-vis-container'>
        <div className='data-vis-header'>
          <p className='data-vis-header-1'>Zonal Annual Mean Land-Ocean Temperature</p>
          <p className='data-vis-header-2'>{ `${ this.props.minDate.getFullYear() } - ${ this.props.maxDate.getFullYear() }` }</p>
        </div>
        <div className='data-vis-body'>
          <div className='data-vis-chart' tabIndex={ -1 } ref={ this.renderChart } />
          <div className='data-vis-tools'>
            {
              this.props.data.map(r =>
                <div key={ r.place } style={ { color: this.props.placeProperties [ r.place ].color } }>
                  <div className='bx--form-item'>
                    <input className='bx--toggle bx--toggle--small' id={ `toggle-${ r.place }` } type='checkbox'
                           aria-label='Label Name'
                           checked={ this.state.selectedPlaces.find(d => d === r.place) !== undefined }
                           onChange={ e => this.togglePlace(r.place, e.target.checked) }
                    />
                    <label className='bx--toggle__label' htmlFor={ `toggle-${ r.place }` }>
                      <span className='bx--toggle__appearance'>
                      </span>
                      { this.props.placeProperties [ r.place ].text }
                    </label>
                  </div>
                </div>
              )
            }
          </div>
        </div>
        <div className='data-vis-footer'>
          <p className='data-vis-footer-1'>GISTEMP Team, 2018: GISS Surface Temperature Analysis (GISTEMP). NASA Goddard
            Institute for Space Studies. Dataset accessed 2018-02-10 at https://data.giss.nasa.gov/gistemp/</p>
          <p className='data-vis-footer-2'>Hansen, J., R. Ruedy, M. Sato, and K. Lo, 2010: Global surface temperature
            change, Rev. Geophys., 48, RG4004, doi:10.1029/2010RG000345</p>
        </div>
      </div>
    );
  }

  private windowResized = () => this.chartState.resize();

  private renderChart = (elem: HTMLElement | null) => {
    if (elem !== null) {
      this.chartState = (() => {

        let clientRect = elem.getBoundingClientRect();
        const margin = { top: 20, right: 40, bottom: 20, left: 40 };

        const $svg = d3.select(elem)
                       .append<SVGSVGElement>('svg')
                       .attr('width', clientRect.width)
                       .attr('height', clientRect.height);

        const startDate = this.props.data[ 0 ].values[ 0 ].date;
        const endDate = this.props.data[ 0 ].values[ this.props.data[ 0 ].values.length - 1 ].date;

        const $gRoot = $svg.append<SVGGElement>('g')
                           .attr('transform', `translate(${ margin.left },${ margin.top })`);

        const x = d3.scaleTime()
                    .domain([ startDate, endDate ]);

        const y = d3.scaleLinear()
                    .domain([ this.props.minValue, this.props.maxValue ])
                    .nice();

        const line = d3.line<TemperatureByDate>()
                       .x(d => x(d.date))
                       .y(d => y(d.temperature));

        const nbXTicks = (nbIntervals: number) => {
          // replace the last automatic tick by the real last one
          const ticks = x.ticks(nbIntervals);
          ticks[ ticks.length - 1 ] = endDate;
          return ticks;
        };

        const xAxis = d3.axisBottom(x);

        $gRoot.append<SVGGElement>('g')
              .attr('class', 'data-vis-axis-x');

        const yFormat = d3.format('.1f');

        const yAxis = d3.axisLeft<number>(y)
                        .tickFormat(d => d === 0 ? '0' : `${ yFormat(d) }Â°`);

        $gRoot.append<SVGGElement>('g')
              .attr('class', 'data-vis-axis-y')
              .call(yAxis);

        $gRoot.append('text')
              .attr('class', 'data-vis-info')
              .attr('dy', -3)
              .text('Annual temperature change compared to the mean of the years from 1951 to 1980');

        return {
          update: () => {

            const width = clientRect.width - margin.left - margin.right;
            const height = clientRect.height - margin.top - margin.bottom;
            const nbIntervals = Math.max(Math.trunc(width / 50), 1);

            $svg.attr('width', clientRect.width)
                .attr('height', clientRect.height);

            x.rangeRound([ 0, width ]);
            y.rangeRound([ height, 0 ]);

            xAxis.tickValues(nbXTicks(nbIntervals));
            yAxis.tickSize(-(width + 20));

            $gRoot.select<SVGGElement>('.data-vis-axis-x')
                  .transition()
                  .attr('transform', `translate(0,${ height })`)
                  .call(xAxis);

            $gRoot.select<SVGGElement>('.data-vis-axis-y')
                  .transition()
                  .call(yAxis)
                  .call($g => {
                    $g.selectAll('.tick')
                      .attr('class', d => `tick ${ d === 0 ? 'data-vis-axis-y-zero' : '' }`)
                      .call($g1 => $g1.select('line')
                                      .attr('x1', -20))
                      .call($g1 => $g1.select('text')
                                      .attr('dx', -3)
                                      .attr('dy', -3));
                  });

            const $chartLines = $gRoot.selectAll<SVGGElement, TemperaturesByPlace>('.data-vis-group-line')
                                      .data(this.props.data.filter(d => this.state.selectedPlaces.find(p => p === d.place) !== undefined), d => d.place);

            $chartLines.exit()
                       .remove();

            $chartLines.enter()
                       .append<SVGGElement>('g')
                       .attr('class', 'data-vis-group-line')
                       .call($g => $g.append<SVGPathElement>('path')
                                     .attr('class', 'data-vis-line')
                                     .style('stroke', d => this.props.placeProperties [ d.place ].color)
                       )
                       .merge($chartLines)
                       .transition()
                       .call($g => $g.select('.data-vis-line')
                                     .attr('d', d => line(d.values))
                       );
          },

          resize: () => {
            clientRect = elem.getBoundingClientRect();
            this.chartState.update();
          }
        };
      })();
      this.chartState.update();
    }
  };

  private togglePlace(place: WorldPlace, checked: boolean) {
    this.setState({
      selectedPlaces: setInArray(this.state.selectedPlaces, p => p === place, checked ? place : undefined)
    });
  }
}
