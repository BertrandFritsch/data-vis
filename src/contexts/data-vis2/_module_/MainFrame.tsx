import * as React from 'react';
import './MainFrame.scss';
import * as d3 from 'd3';
import { RawRecord } from './reducers';

export interface Props {
  data: RawRecord[];
}

interface ChartState {
  update: () => void;
  resize: () => void;
}

export default class MainFrame extends React.PureComponent<Props> {
  private chartState: ChartState = { update: () => undefined, resize: () => undefined };

  componentDidMount() {
    window.addEventListener('resize', this.windowResized);
  }

  componentDidUpdate() {
    this.chartState.update();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.windowResized);
  }

  render() {
    return (
      <div className='data-vis-container'>
        <div className='data-vis-header'>
          <p className='data-vis-header-1'>European Union Population</p>
          <p className='data-vis-header-2'>2017</p>
        </div>
        <div className='data-vis-body'>
          <div className='data-vis-chart' tabIndex={ -1 } ref={ this.renderChart } />
        </div>
        <div className='data-vis-footer'>
          <p className='data-vis-footer-1'>Source: https://ec.europa.eu/eurostat/</p>
          <p className='data-vis-footer-2'>Copyright: https://ec.europa.eu/info/legal-notice_en</p>
        </div>
      </div>
    );
  }

  private windowResized = () => this.chartState.resize();

  private renderChart = (elem: HTMLElement | null) => {
    if (elem !== null) {
      this.chartState = (() => {

        let clientRect = elem.getBoundingClientRect();
        const margin = { top: 20, right: 40, bottom: 20, left: 40 };

        const $svg = d3.select(elem)
                       .append<SVGSVGElement>('svg')
                       .attr('width', clientRect.width)
                       .attr('height', clientRect.height);

        return {
          update: () => {

            const width = clientRect.width - margin.left - margin.right;
            const height = clientRect.height - margin.top - margin.bottom;

            $svg.attr('width', clientRect.width)
                .attr('height', clientRect.height);
          },

          resize: () => {
            clientRect = elem.getBoundingClientRect();
            this.chartState.update();
          }
        };
      })();
      this.chartState.update();
    }
  };
}
