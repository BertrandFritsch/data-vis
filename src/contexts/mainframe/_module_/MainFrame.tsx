import * as React from 'react';
import './MainFrame.scss';
import * as d3 from 'd3';
import { Record } from './reducers';

export interface Props {
  data: Record[];
}

export default class MainFrame extends React.PureComponent<Props> {
  private chartState: { update: () => void } = { update: () => undefined };

  render() {
    return (
      <div className='data-vis-container'>
        <div className='data-vis-header' />
        <div className='data-vis-body' tabIndex={ -1 } ref={ this.renderChart } />
      </div>
    );
  }

  private renderChart = (elem: HTMLElement | null) => {
    if (elem !== null) {
      this.chartState = (() => {

        const $svg = d3.select(elem)
                       .append<SVGSVGElement>('svg');

        const clientRect = $svg.node()!.getBoundingClientRect();

        const margin = { top: 20, right: 10, bottom: 20, left: 10 };
        const width = clientRect.width - margin.left - margin.right;
        const height = clientRect.height - margin.top - margin.bottom;

        const $gRoot = $svg.append<SVGGElement>('g')
                           .attr('transform', `translate(${ margin.left },${ margin.top })`);

        console.log(d3.extent<Record, Date>(this.props.data, d => new Date(d.Year, 0, 1)) as [ Date, Date ])

        const x = d3.scaleTime()
                    .rangeRound([ 0, width ])
                    .domain(d3.extent<Record, Date>(this.props.data, d => new Date(d.Year, 0, 1)) as [ Date, Date ]);

        const y = d3.scaleLinear()
                    .rangeRound([ height, 0 ])
                    .domain(d3.extent<Record, number>(this.props.data, d => d.Glob) as [ number, number ]);

        const line = d3.line<Record>()
                       .x(d => x(new Date(d.Year, 0, 1)))
                       .y(d => y(d.Glob));

        $gRoot.append<SVGGElement>('g')
              .attr('transform', `translate(0,${ height / 2 })`)
              .call(d3.axisBottom(x));
              // .select('.domain')
              // .remove();

        $gRoot.append<SVGGElement>('g')
              .call(d3.axisLeft(y))
              .append('text')
              .attr('fill', '#000')
              .attr('transform', 'rotate(-90)')
              .attr('y', 6)
              .attr('dy', '0.71em')
              .attr('text-anchor', 'end')
              .text('Price ($)');

        $gRoot.append('path')
              .datum(this.props.data)
              .attr('fill', 'none')
              .attr('stroke', 'steelblue')
              .attr('stroke-linejoin', 'round')
              .attr('stroke-linecap', 'round')
              .attr('stroke-width', 1.5)
              .attr('d', line);

        return {
          update: () => {

          }
        };
      })();
      this.chartState.update();
    }
  };
}
