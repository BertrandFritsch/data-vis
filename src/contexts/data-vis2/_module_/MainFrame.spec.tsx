import { mount, shallow } from 'enzyme';
import * as React from 'react';
import MainFrame, { Props } from './MainFrame';

// tslint:disable-next-line no-var-requires
const data = require('../../../../data/data2.json');

describe('<MainFrame />', () => {
  const getInitialProps = (): Props => ({
    data
  });

  let root: HTMLDivElement = document.createElement('div');

  jest.useFakeTimers();

  beforeEach(() => {
    root = document.createElement('div');
    document.body.appendChild(root);
  });

  afterEach(() => {
    jest.runAllTimers();
    root.parentNode!.removeChild(root);
  });

  it('should render correctly', () => {

    const props = { ...getInitialProps() };

    const wrapper = shallow<Props>(<MainFrame { ...props } />);

    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the svg element', () => {

    const props = { ...getInitialProps() };

    mount<Props>(<MainFrame { ...props } />, { attachTo: root });
    jest.runAllTimers();

    expect(root).toMatchSnapshot();
  });

  it('should adapt when the window is resized', () => {

    const props = { ...getInitialProps() };

    const wrapper = mount<Props>(<MainFrame { ...props } />, { attachTo: root });

    root.querySelector('svg')!.getBoundingClientRect = () => ({ top: 0, right: 500, bottom: 500, left: 0, width: 500, height: 500 });
    window.dispatchEvent(new Event('resize'));
    wrapper.update();
    jest.runAllTimers();

    expect(root).toMatchSnapshot();
  });
});
