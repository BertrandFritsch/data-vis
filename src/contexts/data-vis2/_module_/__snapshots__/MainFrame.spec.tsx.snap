// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainFrame /> should adapt when the window is resized 1`] = `
<div>
  <div
    class="data-vis-container"
  >
    <div
      class="data-vis-header"
    >
      <p
        class="data-vis-header-1"
      >
        European Union Population
      </p>
      <p
        class="data-vis-header-2"
      >
        2017
      </p>
    </div>
    <div
      class="data-vis-body"
    >
      <div
        class="data-vis-chart"
        tabindex="-1"
      >
        <svg
          height="1000"
          width="1000"
        />
      </div>
    </div>
    <div
      class="data-vis-footer"
    >
      <p
        class="data-vis-footer-1"
      >
        Source: https://ec.europa.eu/eurostat/
      </p>
      <p
        class="data-vis-footer-2"
      >
        Copyright: https://ec.europa.eu/info/legal-notice_en
      </p>
    </div>
  </div>
</div>
`;

exports[`<MainFrame /> should render correctly 1`] = `
<div
  className="data-vis-container"
>
  <div
    className="data-vis-header"
  >
    <p
      className="data-vis-header-1"
    >
      European Union Population
    </p>
    <p
      className="data-vis-header-2"
    >
      2017
    </p>
  </div>
  <div
    className="data-vis-body"
  >
    <div
      className="data-vis-chart"
      tabIndex={-1}
    />
  </div>
  <div
    className="data-vis-footer"
  >
    <p
      className="data-vis-footer-1"
    >
      Source: https://ec.europa.eu/eurostat/
    </p>
    <p
      className="data-vis-footer-2"
    >
      Copyright: https://ec.europa.eu/info/legal-notice_en
    </p>
  </div>
</div>
`;

exports[`<MainFrame /> should render correctly 2`] = `""`;

exports[`<MainFrame /> should render the svg element 1`] = `
<div>
  <div
    class="data-vis-container"
  >
    <div
      class="data-vis-header"
    >
      <p
        class="data-vis-header-1"
      >
        European Union Population
      </p>
      <p
        class="data-vis-header-2"
      >
        2017
      </p>
    </div>
    <div
      class="data-vis-body"
    >
      <div
        class="data-vis-chart"
        tabindex="-1"
      >
        <svg
          height="1000"
          width="1000"
        />
      </div>
    </div>
    <div
      class="data-vis-footer"
    >
      <p
        class="data-vis-footer-1"
      >
        Source: https://ec.europa.eu/eurostat/
      </p>
      <p
        class="data-vis-footer-2"
      >
        Copyright: https://ec.europa.eu/info/legal-notice_en
      </p>
    </div>
  </div>
</div>
`;
